Feature `CadShape` node

ad-daniel opened the pull request:
**Description**
To address this issue: https://github.com/cyberbotics/webots/issues/4195, this PR creates a full node (previously a PROTO) for the importing of collada objects

- [x] Add fields:
  - [x] url as MFString
  - [x] ccw
  - [x] castShadows
  - [x] isPickable
- [x] add segmentation and range finder visibility
- [x] add downloader
- [x] prevent it from being used in bounding object
- [x] replace pbr tokenizer init by direct init (and fix crash)
- [x] remove colladashape proto related code
- [x] remove model import and VRML export
- [x] add test 
- [x] docs https://cyberbotics.com/doc/reference/colladashape?version=feature-collada-shape-node
- [x] changelog (with backwards compatibility notice ColladaShapes -> ColladaShape)

Following the discussion here https://github.com/cyberbotics/urdf2webots/issues/168, the node should be made more generic and support both wavefront and collada files

- [x] add obj support
- [x] add test
- [x] update docs

omichel commented on src/webots/nodes/WbAbstractAppearance.cpp:
```diff
@@ -37,6 +40,23 @@ WbAbstractAppearance::WbAbstractAppearance(const WbNode &other) : WbBaseNode(oth
+
+WbAbstractAppearance::~WbAbstractAppearance() {
+  delete mName;
+  if (mTextureTransform)
+    delete mTextureTransform;
```
It is OK to call delete with a NULL argument (it will simply do nothing):
```suggestion
  delete mTextureTransform;
```

omichel commented on src/webots/nodes/WbPbrAppearance.cpp:
```diff
@@ -68,6 +70,77 @@ WbPbrAppearance::WbPbrAppearance(const WbNode &other) : WbAbstractAppearance(oth
+  material->Get(AI_MATKEY_COLOR_EMISSIVE, emissiveColor);
+  mEmissiveColor = new WbSFColor(emissiveColor[0], emissiveColor[1], emissiveColor[2]);
+
+  float opacity = 1.0f;
```
```suggestion
  const float opacity = 1.0f;
```

omichel commented on src/webots/nodes/WbColladaShape.cpp:
```diff
@@ -0,0 +1,508 @@
+  aiColor3D emissiveColor(0.0f, 0.0f, 0.0f);
+  material->Get(AI_MATKEY_COLOR_EMISSIVE, emissiveColor);
+
+  float opacity = 1.0f;
```
```suggestion
  const float opacity = 1.0f;
```

omichel commented on src/webots/nodes/WbPbrAppearance.cpp:
```diff
@@ -76,6 +149,31 @@ WbPbrAppearance::~WbPbrAppearance() {
+    delete mOcclusionMapStrength;
+    delete mEmissiveIntensity;
+    // maps
+    if (mBaseColorMap)
+      delete mBaseColorMap;
+    if (mRoughnessMap)
+      delete mRoughnessMap;
+    if (mMetalnessMap)
+      delete mMetalnessMap;
+    if (mNormalMap)
+      delete mNormalMap;
+    if (mOcclusionMap)
+      delete mOcclusionMap;
+    if (mEmissiveColorMap)
+      delete mEmissiveColorMap;
```
No need to test for NULL before deleting objects.

omichel reviewed (changes requested):
I believe the approach is pretty good.
Maybe it would help a little bit to add a comment to the shallow constructors so that we understand that these constructors should be used only in the special case where we want to build a shallow version of the node.

omichel commented on src/webots/vrml/WbSFColor.hpp:
```diff
@@ -27,6 +27,7 @@ class WbSFColor : public WbSingleValue {
   Q_OBJECT
 
 public:
+  explicit WbSFColor(double r, double b, double g) : mValue(WbRgb(r, g, b)){};
```
```suggestion
  explicit WbSFColor(double r, double b, double g) : mValue(WbRgb(r, g, b)) {}
```

omichel commented on src/webots/vrml/WbSFString.hpp:
```diff
@@ -26,6 +26,7 @@ class WbSFString : public WbSingleValue {
   Q_OBJECT
 
 public:
+  explicit WbSFString(const QString &s) : mValue(s){};
```
```suggestion
  explicit WbSFString(const QString &s) : mValue(s) {}
```

omichel commented on src/webots/nodes/WbColladaShape.cpp:
```diff
@@ -0,0 +1,529 @@
+  importer.SetPropertyInteger(AI_CONFIG_PP_RVC_FLAGS, aiComponent_CAMERAS | aiComponent_LIGHTS | aiComponent_BONEWEIGHTS |
+                                                        aiComponent_ANIMATIONS | aiComponent_COLORS);
+
+  unsigned int flags = aiProcess_ValidateDataStructure | aiProcess_Triangulate | aiProcess_GenSmoothNormals |
```
```suggestion
  const unsigned int flags = aiProcess_ValidateDataStructure | aiProcess_Triangulate | aiProcess_GenSmoothNormals |
```

omichel commented on src/webots/nodes/WbColladaShape.cpp:
```diff
@@ -0,0 +1,529 @@
+        continue;
+
+      if (vertices > 100000)
+        warn(tr("mesh '%1' has more than 100'000 vertices, it is recommended to reduce the number of vertices.")
```
```suggestion
        warn(tr("Mesh '%1' has more than 100'000 vertices, it is recommended to reduce the number of vertices.")
```

omichel reviewed (changes requested):
Overall, it looks very good to me.
I will now test it.

omichel reviewed (commented):
There is no example in the distribution using ColladaShape.
It would be nice to either modify an existing proto or world to use ColladaShape or add a new example.
Or maybe it's coming in #4194?

ad-daniel commented on src/webots/nodes/WbAbstractAppearance.cpp:
```diff
@@ -37,6 +39,23 @@ WbAbstractAppearance::WbAbstractAppearance(const WbNode &other) : WbBaseNode(oth
 
+WbAbstractAppearance::WbAbstractAppearance(const QString &modelName, const aiMaterial *material) :
+  WbBaseNode(modelName, material) {
+  aiString name("PBRAppearance");
```
These can't be const since they are passed to assimp which is supposed to replace the contents (only if it finds it)

ad-daniel commented:
Basically there's 2 issues going on:
- On linux the cat model fails to open because assimp cannot decode the collada file (it contains empty tags such as `<subject/>` which it doesn't know how to handle). I believe msys installs assimp 5.2.3 whereas linux/mac install 5.0.1. I've upgraded to 5.2.3 on Linux as well and it seems the issue goes away (and save 70MB in the process, the compiled library is significantly smaller)
- The reason it looked like it wasn't working on windows is because the collada file contains `<transparency>0.0</transparency>` tag. In my implementation, to determine the transparency I check first for the presence of `AI_MATKEY_TRANSPARENCYFACTOR`, and if that fails I determine the transparency from the `AI_MATKEY_OPACITY` value, setting the transparency to `1 - opacity` (as far as I can tell, 90% of the times the latter is what is defined). The issue is that for reasons I don't know, that transparency in the collada file ends up being encoded as an opacity, so I ended up enforcing `1.0 - 0.0 = 1.0` as transparency, making the cat invisible. So I don't think there's an issue with the code, I don't see much else we can do other than report the behavior and hope that at some point it gets fixed, but until then some files might have the same behavior

ad-daniel commented:
The node is now generic and supports both collada and wavefront. A test was added that fails if all of a sudden the mapping changes for either format. It also ensures that a node added using `CadShape` compared to one built manually using `Shape { appearance PBRAppearance {} geometry Mesh {}}` (having the same roughness, basecolor, ... parameters of course) match.
A direct comparison between a `CadShape { url "something.dae" }` and `CadShape { url "same_something.obj" }` was not added on purpose since although similar they hardly ever would match in practice (since for instance obj defines the basecolor from a mat file, whereas collada does it differently).

Concerning the transparency, I've looked in the assimp codebase and it's a bit messy there as well, the reason is that different software programs apply the specifications of the various file formats in a very inconsistent/incomplete way, so assimp attempts to do its best while remaining as close as possible to the official specifications. In particular, I'm not even certain they consider it to be a bug since according to the code, depending on the situation, the transparency is flipped on purpose.
Since querying just the AI_MATKEY_OPACITY has worked well for years (for instance when importing from `3d model import` menu, or more recently with ColladaShapes PROTO), I've reverted my implementation to that approach instead (compared to the approach I used in this PR where I queried several MATKEYs related to transparency, not just AI_MATKEY_OPACITY).

Concerning the issues raised here https://github.com/cyberbotics/webots/issues/4447, the scaling was solved in a PR targetting master and the remaining issue about the inconsistency in the orientation (i.e., exporting as X-forward and on Webots not being X-forward) I will address it on a separate PR since it might entail updating all worlds that rely on the Mesh node.

This PR is currently blocking: https://github.com/cyberbotics/webots/pull/4194 and https://github.com/cyberbotics/urdf2webots/pull/169.

ad-daniel commented:
PS: there is no rush for the review.

omichel commented on docs/reference/cadshape.md:
````diff
@@ -0,0 +1,41 @@
+}
+```
+
+The [CadShape](#cadshape) node renders a Collada object (.dae) or Wavefront (.obj) imported from an external file.
````
```suggestion
The [CadShape](#cadshape) node renders a Collada (.dae) or Wavefront (.obj) object imported from an external file.
```

omichel commented on docs/reference/cadshape.md:
```diff
@@ -0,0 +1,41 @@
+The [CadShape](#cadshape) node renders a Collada object (.dae) or Wavefront (.obj) imported from an external file.
+Both Collada and Wavefront files can include 3D geometries and appearances, be it materials or references to external textures, therefore the [CadShape](#cadshape) node builds both.
+If the file references external textures, these should be placed relative to the collada/wavefront file itself as indicated (i.e., in the location expressed within the .dae/.obj file), no additional search is performed for these textures.
+
```
```suggestion
```

omichel commented on tests/rendering/worlds/cadshape_node.wbt:
```diff
@@ -0,0 +1,344 @@
+  supervisor TRUE
+}
+TestSuiteSupervisor {
+}
```
```suggestion
}

```

omichel reviewed (changes requested):
Since this PR is quite big, it would be nice to have a second review.

BenjaminDeleze reviewed (changes requested):
Seems to works well! However, I have some remarks:

- `CadShape` is missing from `nodes-and-api-functions.md`
- While I had no problem to generate an animation/scene, I did not manage to make the streaming work. It seems to be a problem of textures path.
- Is it deliberate to remove the vrml export feature ? If it is the case, I think we can clean a lot of thing as (if I am not mistaken) the only exporter left is x3d.
- `Move viewpoint to object` does not work (if I remember correctly it could be because of the bounding sphere)

ad-daniel commented on src/webots/nodes/WbPbrAppearance.cpp:
```diff
@@ -595,3 +685,48 @@ bool WbPbrAppearance::exportNodeHeader(WbVrmlWriter &writer) const {
   return WbAbstractAppearance::exportNodeHeader(writer);
 }
+
+void WbPbrAppearance::exportShallowNode(WbVrmlWriter &writer) const {
```
I think it's better yes, because they have been instantiated "artificially" (i.e not from a tokenizer nor copied from an existing node) so most of the member variables of these nodes, for instance at the basenode level, are not defined. Hence although letting the "normal" writer do the export might technically be possible, it entails adding a ton of conditionals just about everywhere to prevent segmentation faults at every corner, which is not really a clean approach and definitely harder to maintain in the long term. Or at least, that was the rationale behind the choice

ad-daniel commented on resources/web/wwi/nodes/WbCadShape.js:
```diff
@@ -0,0 +1,75 @@
+import WbWorld from './WbWorld.js';
+import {getAnId} from './utils/utils.js';
+
+export default class WbCadShape extends WbBaseNode {
```
The reason was that once webotsjs will support tessellation, it will be closer to a WbBaseNode than a WbGroup, but you're right, now it's definitely the opposite

ad-daniel commented on resources/web/wwi/Parser.js:
```diff
@@ -550,6 +553,33 @@ export default class Parser {
+    let url = getNodeAttribute(node, 'url', '');
+    if (typeof url !== 'undefined')
+      url = url.split('"').filter(element => element)[0]; // filter removes empty elements
+    const ccw = parseFloat(getNodeAttribute(node, 'ccw', 'true'));
```
You're right, it's how ccw was encoded for indexedfacesets before the cleanup the recent cleanup of webotsjs and I forgot to update it accordingly on this branch, I'll do it now

BenjaminDeleze commented:
When streaming a world with a `CadShape`, a folder with the textures of the `CadShape` is created. For example, when I stream the test world I got:
`$WEBOTS_HOME/textures/cube.jpg`

Maybe the `CadShape` images are put there instead of being in `localhost:1234/textures` and that is why the steaming viewer is not finding the textures. (Just a hypothesis)

omichel commented on src/webots/nodes/WbSphere.cpp:
```diff
@@ -116,101 +116,20 @@ bool WbSphere::areSizeFieldsVisibleAndNotRegenerator() const {
-    writeExport(writer);
-  else
-    WbGeometry::write(writer);
+  WbGeometry::write(writer);  // TOREM
```
What means  TOREM?

ad-daniel commented:
Apologies, I think there's more that can be cleaned up, I just forgot to put it back to draft

ad-daniel commented:
I think the PR can already be reviewed, the world update test will take a while but locally it came up clean, so fingers crossed should be fine. Linux test suite failed again due to `check_urdf` command missing, which appears to be happening a lot the last few days, or maybe it's just my impression.

I've cleaned up the obvious referenced to vrml exporter. Given that there's multiple writers still (urdf, webots), I don't think many of the `isX3d` can be removed  yet, or at least, not safely since we don't do much, to my knowledge, in terms of testing the exporters. So I've cleaned up the obvious cases only, with several exporters at play and little in terms of tests, it'd be a bit of a gamble to go further at this moment.

ad-daniel commented:
PS: I've also fixed an edge case when parsing .dae as when opening some of the collada files from the franka emika robot some of them triggered an assert.

omichel commented on src/webots/vrml/WbMFDouble.cpp:
```diff
@@ -15,7 +15,7 @@
 #include "WbToken.hpp"
 #include "WbTokenizer.hpp"
-#include "WbVrmlWriter.hpp"
+#include "WbWriter.hpp"
```
This one is likely useless:
```suggestion
```

omichel commented on src/webots/vrml/WbMFInt.cpp:
```diff
@@ -15,7 +15,7 @@
 #include "WbToken.hpp"
 #include "WbTokenizer.hpp"
-#include "WbVrmlWriter.hpp"
+#include "WbWriter.hpp"
```
```suggestion
```

omichel commented on src/webots/vrml/WbProtoList.cpp:
```diff
@@ -21,7 +21,7 @@
 #include "WbStandardPaths.hpp"
 #include "WbTokenizer.hpp"
-#include "WbVrmlWriter.hpp"
+#include "WbWriter.hpp"
```
```suggestion
```

omichel commented on src/webots/vrml/WbProtoModel.cpp:
```diff
@@ -27,7 +27,7 @@
 #include "WbTokenizer.hpp"
 #include "WbValue.hpp"
-#include "WbVrmlWriter.hpp"
+#include "WbWriter.hpp"
```
```suggestion
```

omichel commented on src/webots/vrml/WbSingleValue.cpp:
```diff
@@ -13,7 +13,7 @@
 
 #include "WbSingleValue.hpp"
-#include "WbVrmlWriter.hpp"
+#include "WbWriter.hpp"
```
```suggestion
```

BenjaminDeleze reviewed (approved):
Thank you
